### PR TITLE
SilverStripe 4 upgrade

### PR DESCRIPTION
**Pull request by scott-nz**
Upgrade module to be compatible with SilverStripe 4.
HtmlEditorField functionality is removed as the same functionality is available in core SilverStripe

- [x] AC1: All `UploadField` functionality remains.
- [x] AC2: CMS user is able to select an existing directory to upload files into.
- [x] AC3: CMS user selected upload directory is remembered and will apply to all `SelectUploadField` inputs until Session expires
- [x] AC4: Developer can set a default upload folder. CMS users retain ability to select a different folder
- [x] AC5: Developer can restrict ability to select upload destination based on user permissions `setCanSelectFolder('ADMIN')`
- [x] AC6: Developer can disable folder selection for all users `setCanSelectFolder(false)`
- [x] AC7: All instances of `UploadField` can be replaced with `SelectUploadField` using `SilverStripe\Core\Injector`

### CODE REVIEW

**Comment by wilr** [comment]
@scott-nz Any chance this can get merged and `2.0` tagged asap?

**Comment by NightJar** [comment]
Could you please indicate if this is ready by assessing each AC in your OP and checking the ones this work meets?
I assume all, but as they're unchecked checkboxes it presents as if this PR is not ready to be merged.

File: src/SelectUploadField.php [hunk 1]
```diff
@@ -0,0 +1,282 @@
...
+        return parent::getFolderName();
+    }
+
+    /**
+     * @return null|string|string[]
+     */
+    public function getDisplayFolderName()
+    {
+        $name = $this->getFolderName();
+        return preg_replace('/\s*\\/\s*/', ' / ', trim($name, '/'));
```
[the remark this answers is not in the document]
Comment (scott-nz) [reply src/SelectUploadField.php]: comment added describing functionality

File: src/FolderDropdownField.php [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+        parent::__construct($name, $title, $sourceObject, $keyField, $labelField, $showSearch);
+        $this->setValue(self::get_last_folder());
+    }
+
+    /**
+     * Set the last folder selected
+     *
+     * @param int|Folder $folder Folder instance or ID
+     */
+    public static function set_last_folder($folder)
```
[the remark this answers is not in the document]
Comment (scott-nz) [reply src/FolderDropdownField.php]: removed static declaration. 
Function only ever gets passed an ID. Changed function names to reflect actual functionality, removed unneeded complexity and added Int type hint

File: src/FolderDropdownField.php [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+        parent::__construct($name, $title, $sourceObject, $keyField, $labelField, $showSearch);
+        $this->setValue($this->getLastFolderID());
+    }
+
+    /**
+     * Set the last folder selected
+     *
+     * @param int $folderID Folder ID
+     */
+    public function setLastFolderID(Int $folderID)
```
[the remark this answers is not in the document]
Comment (ScopeyNZ) [reply src/FolderDropdownField.php]: We technically still support 5.6 so we should have scalar type hints at all.

File: tests/SelectUploadFieldTestForm.php [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+    public function getRecord()
+    {
+        if (empty($this->record)) {
+            $this->record = DataObject::get_one(SelectUploadFieldTestRecord::class, '"Title" = \'Record1\'');
+        }
+        return $this->record;
+    }
+
+    public function __construct($controller = null, $name = 'Form')
```
[the remark this answers is not in the document]
Comment (NightJar) [reply tests/SelectUploadFieldTestForm.php]: Change made to docblock, not this line. Marking resolved.

File: src/SelectUploadField.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (NightJar) [reply src/SelectUploadField.php]: Comment added to docblock, not this line. Marking as resolved.

File: src/FolderDropdownField.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (NightJar) [reply src/FolderDropdownField.php]: Hmm, a fair point. @scott-nz could you please re-update? Sorry :/

[the remark this answers is not in the document]
Comment (NightJar) [reply src/FolderDropdownField.php]: ... Oh, you already did 😅

File: templates/SilverStripe/SelectUpload/SelectUploadField.ss [hunk 5]
```diff
@@ -4,20 +4,20 @@
             <div class="ss-uploadfield-item-info">
                 <label class="ss-uploadfield-item-name">
                     <% if $CanSelectFolder %>
-                        <small><% _t('UploadField.ADDTO', 'Add to') %>
+                        <small><%t SilverStripe\\SelectUploadField.ADDTO 'Add to' %>
```
Comment (NightJar) [inline templates/SilverStripe/SelectUpload/SelectUploadField.ss]: Can this please match the class name?
`SilverStripe\\SelectUpload\\SelectUploadField.ADDTO`?

Same for all other 4 :)

Comment (scott-nz) [reply templates/SilverStripe/SelectUpload/SelectUploadField.ss]: done :)

**Comment by NightJar** [comment]
@wilr if you can confirm there are no issues with your use-case against `dev-master`, then I'll tag 2.0.0 :)

**Comment by NightJar** [comment]
The team having done the work has tested it, I'm sure.
However testing for a single use case is limited, I'd prefer more feedback. I agree a beta is good for now though.

**Comment by wilr** [comment]
Thanks all for the hard work and the release 🍻

**Pull request merged**